### PR TITLE
Set the stdout_path and stderr_path the job/task should write to

### PR DESCRIPTION
**Pull request by WilliamMcCumstie**
The stdout and stderr paths are optional inputs to job creation. They direct the daemon to write the job/task output to the specified location.

They support various replacements that can customise the paths on a per node/job basis. See the documentation for a full list of supported replacements. Note the description is in GFM.
```
curl http://<ip>/v0/docs | jq '.definitions.newJob.properties.attributes.properties["stdout-path"].description'
curl http://<ip>/v0/docs | jq '.definitions.newJob.properties.attributes.properties["stderr-path"].description'
```

This PR is based on #12

### CODE REVIEW

File: lib/flight_scheduler/path_generator.rb [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+      'a' => 'The current index when running an array task, otherwise 0',
+      # TODO: This can't be implemented affectively as the nodes are allocated
+      # to individual array tasks and not the job. This makes determining the
+      # node indexing either inconsistent or impossible to determine
+      #
+      # The allocation behaviour is likely to change TBD. Consider revisiting
+      # 'n' => 'The relative ID of the node within the job',
+    }
+    ALPHA_KEYS = {
+      'A' => "The ID of the associate job when running an array task, otherwise empty string",
```
Comment (benarmston) [inline lib/flight_scheduler/path_generator.rb]: Typo: `associated`.

File: lib/flight_scheduler/path_generator.rb [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+      'j' => 'The ID of non array task jobs, otherwise empty string',
```
Comment (benarmston) [inline lib/flight_scheduler/path_generator.rb]: This should be defined for array tasks too.  It should expand to the array tasks id.

File: lib/flight_scheduler/path_generator.rb [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+    DESC = [
+      'The paths may contain various special characters which will be replaced:',
+      '',
+      *NUMERIC_KEYS.map { |k, d| " * `%\\d*#{k}`: #{d}" },
+      *ALPHA_KEYS.map { |k, d| " * `%#{k}`: #{d}" },
+      " * `%%`: Escape a literal percent character '%', instead of a special directive",
+      ' * `%<char>`: All other characters form an invalid replacement'
+    ].join("\n")
+
+    ALL_CHARS = [*ALPHA_KEYS.keys, *NUMERIC_KEYS.keys.join('')]
```
Comment (benarmston) [inline lib/flight_scheduler/path_generator.rb]: What's the purpose of the `join('')` on `NUMERIC_KEYS`.  We don't have one on `ALPHA_KEYS` and we do have one later on `ALL_CHARS`.

File: lib/flight_scheduler/path_generator.rb [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+    attr_reader :node, :job, :task
+
+    def initialize(node:, job:, task: nil)
+      @node = node
+      @job = job
+      @task = task
+    end
+
+    def pct_a
+      task ? task.array_index : 0
```
Comment (benarmston) [inline lib/flight_scheduler/path_generator.rb:76]: It looks like slurm uses `2**32 - 2` for non-array jobs here.  I'm not sure how consistently it uses that number across platforms and architectures, but it is consistent on the single machine I've tested it.

I can't think of any good reason for us to copy that.  `0` seems like a suitably valid value, and I think it unlikely that anyone is relying on `%a` expanding to `2**32 - 2` for non-array jobs. Probably.

File: lib/flight_scheduler/path_generator.rb [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+    end
+
+    def pct_A
+      task ? job.id : ''
```
Comment (benarmston) [inline lib/flight_scheduler/path_generator.rb]: On further inspection of what slurm does, for non-array tasks this should expand to the job's id.

File: lib/flight_scheduler/path_generator.rb [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+    end
+
+    def pct_j
+      task ? '' : job.id
```
Comment (benarmston) [inline lib/flight_scheduler/path_generator.rb]: For array tasks, this should expand to the array task's id.

**Review by benarmston — CHANGES_REQUESTED** [review]
@WilliamMcCumstie this is a good start. I've left a few comments in line for you to look at and there are a couple of general issues to mention.

There is a bug with how standard error is handled.  If standard error and standard output are both set to the same file and a submission script writes to both standard out and standard error one of those is lost.

The reason is that the file is that `::Process.spawn(..., out: 'some/path', err: 'some/path')` will open `some/path` twice if both standard output and standard error are written to.  Both of the following should work correctly.

```ruby
::Process.spawn(..., [:out, :err] => 'some/path')
::Process.spawn(..., out: 'some/path', :err => :out)
```

---

There are also a number of inconsistencies with how slurm handles the path patterns.  These were not mentioned on the trello card as I didn't know about them at the time :-)

* `%` should render as `%`
* `%%%` should render as `%%`
* unsupported formatting characters should render litteraly, e.g., `%q` renders as `%q`.
* Non-numeric formats "support" padding.  E.g., `%5N` should render exactly as `%N` renders.

Ideally, we'd be consistent with slurm on our rendering of these.  However, I'm content to introduce inconsistency if it greatly simplifies the implementation of time to fix.

As far as I can tell, slurm never rejects a job due to an invalid path pattern.  It just might not render as you would expect.

File: lib/flight_scheduler/path_generator.rb (hunk 4 above)
Comment (WilliamMcCumstie) [reply lib/flight_scheduler/path_generator.rb:76]: Hmm `2**32 -2` looks like an odd number. My guess it's "-1" as an unsigned 32-bit int

Comment (WilliamMcCumstie) [reply lib/flight_scheduler/path_generator.rb:76]: Yep! It's an algorithmic underflow :laughing:
```
2**32-2 = 0xFFFFFFFE
```

**Comment by WilliamMcCumstie** [comment]
Rebased on #18 To resolve various conflicts between the branches. Also allows the `username` to be integrated into the paths

File: lib/flight_scheduler/path_generator.rb (hunk 1 above)
Comment (WilliamMcCumstie) [reply lib/flight_scheduler/path_generator.rb]: https://github.com/openflighthpc/flight-scheduler-controller/pull/15/commits/ae036d1203aa02b303c8571c05d073641644ae47

File: lib/flight_scheduler/path_generator.rb (hunk 3 above)
Comment (WilliamMcCumstie) [reply lib/flight_scheduler/path_generator.rb]: https://github.com/openflighthpc/flight-scheduler-controller/pull/15/commits/ae036d1203aa02b303c8571c05d073641644ae47

File: lib/flight_scheduler/path_generator.rb (hunk 2 above)
Comment (WilliamMcCumstie) [reply lib/flight_scheduler/path_generator.rb]: https://github.com/openflighthpc/flight-scheduler-controller/pull/15/commits/7101482abd793a47918f45ce1046a6fc9b8b2284

File: lib/flight_scheduler/path_generator.rb (hunk 5 above)
Comment (WilliamMcCumstie) [reply lib/flight_scheduler/path_generator.rb]: https://github.com/openflighthpc/flight-scheduler-controller/pull/15/commits/7101482abd793a47918f45ce1046a6fc9b8b2284

File: lib/flight_scheduler/path_generator.rb (hunk 6 above)
Comment (WilliamMcCumstie) [reply lib/flight_scheduler/path_generator.rb]: https://github.com/openflighthpc/flight-scheduler-controller/pull/15/commits/7101482abd793a47918f45ce1046a6fc9b8b2284

**Comment by WilliamMcCumstie** [comment]
I've updated the replacement so all inputs are now valid. This does mean the `valid?` method needed removing (as it would always return true).

I've also added a few more specs for the specific use cases you mentioned.

**Comment by benarmston** [comment]
I've added a commit to keep the autoload list ordered.  Having the list ordered reduces the chances of merge conflicts compared
to having all new entries added to the end of the list.

With that in place this looks good to me.

**Pull request merged**